File: C++/N-Queens.cpp
#include<iostream>
#include<vector>
#include<string>
#include<string.h>
#include<math.h>
using namespace std;
class Solution {
public:
	vector<vector<string> > ret;
	bool isValid(int row, int col, vector<int>& board)
	{
		for(int i = 0; i < row; i++)
		{
			int data = board[i];
			if(data == col || (row+col) == (i+data) || (row-col) == (i-data))
				return false;
		}
		return true;
	}
	void place(int index,int n, vector<int>& board)
	{
		if(index == n)
		{
			vector<string> sol;
			for(int i = 0; i < n; i++)
			{
				string row(n,'.');
				row[board[i]] = 'Q';
				sol.push_back(row);
			}
			ret.push_back(sol);
			return;
		}
		for(int i = 0; i < n; i++)
		{
			if(isValid(index, i, board))
			{
				board[index] = i;
				place(index+1, n, board);
			}
		}
	}
    vector<vector<string> > solveNQueens(int n) {
		ret.clear();
        if(!n)
			return ret;
		vector<int> board(n,0);
		place(0, n, board);
		return ret;
    }
};
int main()
{
	int n = 1;
	Solution sol;
	sol.solveNQueens(n);
	getchar();
}

File: JAVA/N_Queens.java
import java.util.ArrayList;

class Solution {
	ArrayList<String[]> ret = new ArrayList<String[]>();
	int[] pos;
    public ArrayList<String[]> solveNQueens(int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ret.clear();
        if(n == 0)
        	return ret;
        pos = new int[n]; 
        place(0, n);
        return ret;
    }
	private void place(int index, int n) {
		if(index == n)
		{
			String[] str = new String[n];
			for(int i = 0; i < n; i++)
			{
				str[i] = new String();
				for(int j = 0; j < n ;j++)
				{
					if(j == pos[i])
						str[i] += 'Q';
					else
						str[i] += '.';
				}
			}
			ret.add(str);
			return;
		}
		for(int i = 0; i < n; i++)
		{
			if(canPlace(index, i))
			{
				pos[index] = i;
				place(index+1, n);
			}
		}
	}
	private boolean canPlace(int row, int col) {
		for(int i = 0; i < row; i++)
		{
			int coli = pos[i];
			if(coli == col)
				return false;
			if((i+coli) == (row+col))
				return false;
			if((i-coli) == (row-col))
				return false;
		}
		return true;
		
	}
}
public class N_Queens {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		Solution sol = new Solution();
		System.out.println(sol.solveNQueens(1));
	}

}
